src/hotspot/share/oops/resolvedIndyEntry.hpp

```diff
@@ -0,0 +1,140 @@
+  Method* _method;               // Adapter method for indy call
+  u2 _resolved_references_index; // Index of resolved references array that holds the appendix oop
+  u2 _cpool_index;               // Constant pool index
+  u2 _number_of_parameters;      // Number of arguments for adapter method
+  u1 _return_type;               // Adapter method return type
+  u1 _flags;                     // Flags: [0000|00|has_appendix|resolution_failed]
+
+public:
+  ResolvedIndyEntry() :
+    _method(nullptr),
+    _resolved_references_index(0),
+    _cpool_index(0),
+    _number_of_parameters(0),
+    _return_type(0),
+    _flags(0) {}
+  ResolvedIndyEntry(u2 resolved_references_index, u2 cpool_index) :
+    _method(nullptr),
+    _resolved_references_index(resolved_references_index),
+    _cpool_index(cpool_index),
+    _number_of_parameters(0),
+    _return_type(0),
+    _flags(0) {}
+
+  // Bit shift to get flags
+  // Note: Only one flag exists at the moment but more could be added
+  enum {
+      has_appendix_shift        = 1,
+  };
+
+  // Getters
+  Method* method()               const { return _method;                    }
+  u2 resolved_references_index() const { return _resolved_references_index; }
+  u2 constant_pool_index()       const { return _cpool_index;               }
+  u2 num_parameters()            const { return _number_of_parameters;      }
+  u1 return_type()               const { return _return_type;               }
+  bool is_resolved()             const { return _method != nullptr;         }
+  bool has_appendix()            const { return (_flags & (1 << has_appendix_shift))        != 0; }
+  bool resolution_failed()       const { return (_flags & 1)                                != 0; }
+  bool is_vfinal()               const { return false;                                            }
+  bool is_final()                const { return false;                                            }
+  bool has_local_signature()     const { return true;                                             }
+
+  // Printing
+  void print_on(outputStream* st) const;
+
+  // Initialize with fields available before resolution
+  void init(u2 resolved_references_index, u2 cpool_index) {
+    _resolved_references_index = resolved_references_index;
+    _cpool_index = cpool_index;
+  }
+
+  void set_num_parameters(int value) {
+    assert(_number_of_parameters == 0 || _number_of_parameters == value,
+      "size must not change: parameter_size=%d, value=%d", _number_of_parameters, value);
+    Atomic::store(&_number_of_parameters, (u2)value);
+    guarantee(_number_of_parameters == value,
+      "size must not change: parameter_size=%d, value=%d", _number_of_parameters, value);
+  }
+
+  // Populate structure with resolution information
+  void fill_in(Method* m, u2 num_params, u1 return_type, bool has_appendix) {
+    set_num_parameters(num_params);
+    _return_type = return_type;
+    set_flags(has_appendix);
+    Atomic::release_store(&_method, m);
```

Comment:
Add a comment like // set this last.  The method is read lock free from the entry and if set, indicates the rest of the resolution information is valid.